releases_and_versioning/MAVEN-RELEASE-PLUGIN.md

```diff
@@ -0,0 +1,102 @@
+
+There is one issue with the prepare step that we have to deal with first.
+
+### Command-Line GPG Passphrase Entry
+
+Long term, you probably want to set the GPG passphrase automatically instead of through the command line. Instructions for that are here: http://maven.apache.org/plugins/maven-gpg-plugin/usage.html
+
+But if you use this command-line approach, you may encounter the following error:
+
+```"gpg: signing failed: Inappropriate ioctl for device".```
+
+This is because some part of the process is trying to open an interactive window and doesn't have permission:
+
+https://issues.apache.org/jira/browse/MGPG-59?attachmentOrder=asc
+
+Note that this does not happen with all versions of gpg, but the latest version from Homebrew throws this error. Before running the prepare step, run the following command as advised in the ticket:
+
+```gpg --use-agent --armor --detach-sign --output $(mktemp) pom.xml```
+
+This essentially signs the pom so that the later interactive step can be skipped. You may need to do this each time before running "prepare".
+
+## Release
+
+And now finally, you should be able to run the release:prepare step. A few options are necessary. The flag autoVersionSubmodules will do the work of setting all the versions the same in all the poms, so if you have modules in your project use this. The prepare step also builds everything for release and packages the javadoc. If there are errors in the javadoc, this can stop the release from happening. Fix these if you can, but if not another flag will disable this part of the build.
+
+```mvn release:prepare -DautoVersionSubmodules=true -Darguments="-Dgpg.passphrase=<passphrase> -Dmaven.javadoc.skip=true"```
+
+You'll have three interactive choices to make for versions. Maven provides sane defaults, and if they look good, you can just hit enter:
+
+```
+What is the release version for "org.monarchinitiative.phenol:phenol"? (org.monarchinitiative.phenol:phenol) 1.2.5: : 1.2.5
+What is SCM release tag or label for "org.monarchinitiative.phenol:phenol"? (org.monarchinitiative.phenol:phenol) phenol-1.2.5: : v1.2.5
+What is the new development version for "org.monarchinitiative.phenol:phenol"? (org.monarchinitiative.phenol:phenol) 1.2.6-SNAPSHOT: 
+```
+
+The prepare step pushes changes to the remote repository. Once it's finished you will have a tag and release named v1.2.5. All of the poms in that tag will have a version of 1.2.5. The master branch poms will be versioned at 1.2.6-SNAPSHOT.
+
+## Cleanup
+
+The prepare step creates a couple of new files in the main directory - release.properties and pom.xml.releaseBackup. You can set your repository to ignore these in .gitignore to avoid committing them. Or you can clean them up by running:
+
+
+```mvn release:clean```
+
+It appears that `mvn release:perform` essentially does the same thing, but we have not used this.
```

Comment:
I'm curious about this. I'm really surprised that we don't have to use `perform`. We don't need to change it, but I want to play with it too.

Reply:
From the documentation I read, it seems like all perform does is run the changes to make sure they work and then clean up. If so, we probably should just use that instead of preparing. I originally though release:perform did other deployment steps if you gave it server information, but it doesn't seem to.